_posts/2016-12-19-anim_compression_signal_processing.md

```diff
@@ -1,111 +1,96 @@
+
+The scale of output coefficients is a function of the scale and smoothness of our input values. As such it makes sense to perform [range reduction](http://nfrechette.github.io/2016/11/09/anim_compression_range_reduction/) and to normalize our input values.
+
+# Quantization
+After applying the wavelet transform the number of output values will match the number of input values. No compression has happened yet.
+
+As mentioned previously our output values will be partitioned into sub-bands, and a single scale value somewhat centered around zero — both positive and negative. Each sub-band will end up with a different range of values. Larger sub-bands resulting from the first applications of the wavelet function will be filled with high-frequency information while the smaller sub-bands will comprise the low-frequency information. This is important. It means that a single low-frequency coefficient will impact a larger range of values after performing the inverse wavelet transform. Because of this low-frequency coefficients need higher accuracy than high-frequency coefficients.
+
+To achieve compression we will quantize our coefficients into a reduced number of bits while keeping the single scale value with full precision. Due to the nature of the data, we will perform range reduction per sub-band and normalize our values between `[-1.0, 1.0]`. We only need to keep the range extent for reconstruction and simply assume that the range is centered around zero. Quantization might not make sense for the lower frequency sub-bands with 1, 2, 4 coefficients due to the extra overhead of the range extent. Once our values are normalized we can quantize them. To choose how many bits to use per coefficient we can simply hard code a high number such as 16 bits, 12 bits, or alternatively experiment with values in an attempt to optimize a solution to meet an error threshold. Quantization could also be performed globally to reduce the range of information overhead instead of per sub-band depending on the number of input values being processed. For example processing 16 keys at a time.
+
+# Entropy Coding
+In order to be competitive with other techniques, we need to push compression further using [entropy coding](https://en.wikipedia.org/wiki/Entropy_encoding) which is an entirely lossless compression step.
+
+After quantization we obtain a number of integer values on a single scale all centered around zero. The most obvious thing that we can compress now is the fact that we have very few large values. To leverage this we apply a [zigzag transform](https://wiki.multimedia.cx/index.php/Zigzag_Reordering) on our data, mapping negative integers to positive unsigned integers such that values closest to zero remain closest to zero. This transforms our data in such a way that we still end up with very few large values which are significant because it means that most of our values represented in memory now have many leading zeroes.
```

Comment:
"integer values on a single scale all centered around zero" -> "integer values all centered around zero and a single scale"